#Credit Card Fraud Detection Using Logistic Regression

The project entails the analysis of a credit card transaction dataset for fraud detection. The dataset is highly imbalanced, and we use logistic regression to classify the transactions into fraudulent or legitimate.

---
### **About the Dataset** (information published by the dataset collector)
The dataset consists of credit card transactions made by European cardholders in September 2013. The following are some facts about the dataset:
- 284,807 transactions over two days.
- Only 492 transactions (0.172%) are fraudulent.
The feature's type is numerical scaled with PCA except 'Time' and 'Amount'.
The Target Variable:
The 'Class':
  • 0 -> Valid transaction
  • 1 -> Fraudulent transaction

---
### **Goal**
The task will be to work out a logisitic regression based classifier of fraudulent cases. And afterwards perform the relevant evaluations.
### **Contributor/Acknowledgements**
The dataset is shared from various research collaborations between Worldline and Machine Learning Group of ULB. References are to be cited for information. Refer the ones provided below.
---


In [61]:
#importing Libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

## **1. Loading and Exploring the Dataset**
Here, we load the dataset and explore its structure, including basic information, checking for missing values, and understanding class distribution.


In [62]:
#import necessary libraries
import os
import pandas as pd

# check if the dataset is already downloaded
dataset_path = 'creditcard.csv'
if not os.path.exists(dataset_path):
    #download the dataset using Kaggle API
    os.system('kaggle datasets download -d mlg-ulb/creditcardfraud --unzip')

# load the dataset
credit_card_data = pd.read_csv(dataset_path)

# display the first 5 rows
credit_card_data.head()


Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [63]:
#displaying the last 5 rows
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [64]:
#displaying the last 5 rows
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [65]:
#Dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [66]:
#checking for missing values
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [67]:
#class distribution (0 = Legitimate, 1 = Fraudulent)
credit_card_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

## **2. Data Analysis**
We separate the dataset into legitimate and fraudulent transactions and compare statistical measures for these two groups.


In [68]:
#separating legitimate and fraudulent transactions
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

# displaying their shapes
print("Legitimate transactions:", legit.shape)
print("Fraudulent transactions:", fraud.shape)

Legitimate transactions: (284315, 31)
Fraudulent transactions: (492, 31)


In [69]:
# statistical summary of 'Amount' for legitimate transactions
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [70]:
#statistical summary of 'Amount' for fraudulent transactions
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [71]:
#comparing mean values for both classes
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

## **3. Balancing the Dataset**
The dataset is highly imbalanced. To address this, we undersample the legitimate transactions to match the number of fraudulent ones.


In [72]:
#randomly sample legitimate transactions
legit_sample = legit.sample(n=492)

# combine the balanced legitimate transactions with fraudulent ones
new_dataset = pd.concat([legit_sample, fraud], axis=0)

#verifying the new dataset
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

## **4. Splitting the Dataset**
The dataset is split into training and testing sets (80-20 split), maintaining class balance using stratified sampling.


In [73]:
# splitting features (X) and target (Y)
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

# splitting into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

#displaying shapes
print("Original data shape:", X.shape)
print("Training set shape:", X_train.shape)
print("Testing set shape:", X_test.shape)

Original data shape: (984, 30)
Training set shape: (787, 30)
Testing set shape: (197, 30)


## **5. Model Training and Evaluation**
We train a logistic regression model and evaluate its performance on both the training and test datasets.


In [74]:
#initializing the Logistic Regression model
model = LogisticRegression()

#training the model
model.fit(X_train, Y_train)

#making predictions on the training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(Y_train, X_train_prediction)

print("Accuracy on Training Data:", training_data_accuracy)

Accuracy on Training Data: 0.9428208386277002


In [75]:
#making predictions on the test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(Y_test, X_test_prediction)

print("Accuracy on Test Data:", test_data_accuracy)

Accuracy on Test Data: 0.9289340101522843


## **6. Results**
- **Training Accuracy**: 94.28%
- **Test Accuracy:** 92.89%

Despite the class imbalance, the logistic regression model still shows quite decent performance. Further actions could include more complex models or advanced techniques like precision-recall curves, feature engineering, and hyperparameter optimization, although the results are quite promising.


##Credits
This mini project is based on a public available dataset related to credit card fraud detection. The underlying dataset and background has been provided by:
* **Worldline** and the **Machine Learning Group of ULB** (found on kaggle, see https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud?resource=download). Realised by Mohamed ZOUAD.
---
